odoo/addons/base/models/ir_model.py

```diff
@@ -71,6 +101,31 @@ def query_update(cr, table, values, selectors):
     return [row[0] for row in cr.fetchall()]
 
 
+def upsert(cr, table, cols, rows, conflict):
+    """ Insert or update the table with the given rows.
+
+    :param cr: database cursor
+    :param table: table name
+    :param cols: list of column names
+    :param rows: list of tuples, where each tuple value corresponds to a column name
+    :param conflict: list of column names to put into the ON CONFLICT clause
+    :return: the ids of the inserted or updated rows
+    """
+    query = """
+        INSERT INTO {table} ({cols}) VALUES {rows}
+        ON CONFLICT ({conflict}) DO UPDATE SET ({cols}) = ({excluded})
+        RETURNING id
+    """.format(
+        table=quote(table),
+        cols=", ".join(quote(col) for col in cols),
+        rows=", ".join("%s" for row in rows),
+        conflict=", ".join(conflict),
+        excluded=", ".join("EXCLUDED." + quote(col) for col in cols),
+    )
+    cr.execute(query, rows)
```

Comment:
The execute_values helper from psycopg2 might be a bit less ad-hoc? Still assuming 2.7 is a hard requirement.